src/pybind/mgr/cli_api_commands/module.py

```diff
@@ -0,0 +1,233 @@
+from threading import Event
+from typing import Any, Callable, Dict, List
+
+from mgr_module import MgrModule, HandleCommandResult, CLICommand, API
+
+logger = logging.getLogger()
+
+
+class CephCommander:
+    '''
+    Utility class to inspect Python functions and generate corresponding
+    CephCommand signatures (see src/mon/MonCommand.h for details)
+    '''
+    def __init__(self, func: Callable):
+        self.func = func
+        self.signature = inspect.signature(func)
+        self.params = self.signature.parameters
+
+    def to_ceph_signature(self):
+        '''
+        Generate CephCommand signature (dict-like)
+        '''
+        return {
+            'prefix': f'mgr cli {self.func.__name__}',
+            'perm': API.perm.get(self.func)
+        }
+
+    def to_ceph_benchmark(self):
+        '''
+        Generate CephCommand signature (dict-like)
+        '''
+        return {
+            'prefix': f'mgr cli benchmark {self.func.__name__}',
+            'perm': API.perm.get(self.func)
+        }
+
+
+class ThreadedBenchmarkRunner:
+    def __init__(self, func_name: str, data: list, number_of_total_calls: int, number_of_parallel_calls: int):
+        self._func_name = func_name
+        self._data = data
+        self._number_of_total_calls = number_of_total_calls
+        self._number_of_parallel_calls = number_of_parallel_calls
+        self._threads = []
+        self._jobs: Queue = Queue()
+        self.time_accumlated = 0.0
+        self._time = []
+        self._lock = threading.Lock()
+        self._mgr = None
+
+    def set_mgr(self, mgr):
+        self._mgr = mgr
+
+    def benchmark(self):
+        self.start(self._func_name, self._data)
+        self.join()
+        stats = self.get_stats()
+        return HandleCommandResult(0, json.dumps(stats), "Error")
+
+    def start(self, func, args):
+        self.args = args
+        if(self._number_of_total_calls and self._number_of_parallel_calls):
+            for thread_id in range(self._number_of_parallel_calls):
+                new_thread = threading.Thread(target=self.timer,
+                                              args=(self._jobs, func,))
+                self._threads.append(new_thread)
+            for job_id in range(self._number_of_total_calls):
+                self._jobs.put(job_id)
+            for thread in self._threads:
+                thread.start()
+        else:
+            raise BenchmarkException("Number of Total and number of parallel calls must be greater than 0")
+
+    def join(self):
+        for thread in self._threads:
+            thread.join()
+
+    def get_stats(self):
+        stats = {
+            "avg": (self.time_accumlated / self._number_of_total_calls),
+            "min": min(self._time),
+            "max": max(self._time)
+        }
+        return stats
+
+    def timer(self, jobs, func):
+        self._lock.acquire()
+        while not self._jobs.empty():
+            jobs.get()
+            time_start = time.time()
```

Comment:
BTW we should change this timer, based on [this report](https://www.webucator.com/article/python-clocks-explained/), `time.time()` is not precise enough for measuring calls that lie in the millisecond scale, so we should rather use `time.perf_counter()` instead.

What you can do is adding a `get_time =  time.perf_counter` in the beginning of the file and use that `get_time()` instead.

~Additionally, as we are adding lots of individual deltas (which could cause floating point rounding errors), let's use `time.perf_counter_ns()` which returns an integer in nanoseconds.~ (`perf_counter_ns` is [not supported in Py 3.6](https://docs.python.org/3/library/time.html#time.perf_counter_ns)).